core/celo_backend.go

```diff
@@ -43,8 +47,34 @@ func (b *CeloBackend) CallContract(ctx context.Context, call ethereum.CallMsg, b
 	txCtx := vm.TxContext{}
 	vmConfig := vm.Config{}
 
-	evm := vm.NewEVM(blockCtx, txCtx, b.state, b.chainConfig, vmConfig)
+	evm := vm.NewEVM(blockCtx, txCtx, b.State, b.ChainConfig, vmConfig)
 	ret, _, err := evm.StaticCall(vm.AccountRef(evm.Origin), *call.To, call.Data, call.Gas)
 
 	return ret, err
 }
+
+// GetBalanceERC20 returns an account's balance on a given ERC20 currency
+func (b *CeloBackend) GetBalanceERC20(accountOwner common.Address, contractAddress common.Address) (result *big.Int, err error) {
+	token, err := abigen.NewFeeCurrencyCaller(contractAddress, b)
+	if err != nil {
+		return nil, fmt.Errorf("failed to access FeeCurrency: %w", err)
+	}
+
+	balance, err := token.BalanceOf(&bind.CallOpts{}, accountOwner)
+	if err != nil {
+		return nil, err
+	}
+
+	return balance, nil
+}
+
+func GetBalanceOf(config *params.ChainConfig, state *state.StateDB, account common.Address, feeCurrency *common.Address) *big.Int {
```

Comment:
removed the struct

Reply:
Since GetBalanceERC20 is used from the state_transition, I would dare not change it. And we need to check if the feeCurrency is for the native currency which now I removed GetBalanceOf, but added a CeloBackend.GetFeeBalance(account, feeCurrency), which is directly called from the pool.celoBackend of legacypool and blobpool.